src/contracts/wrappers/ERC721FloorWrapper.sol

```diff
@@ -0,0 +1,213 @@
+        }
+        DepositRequestObj memory obj;
+        (obj) = abi.decode(_data, (DepositRequestObj));
+        if (obj.recipient == address(0)) {
+            // Don't allow deposits to the zero address
+            revert InvalidWithdrawRequest();
+        }
+
+        uint256[] memory tokenIds = new uint256[](1);
+        tokenIds[0] = _tokenId;
+        emit TokensDeposited(tokenIds);
+
+        _mint(obj.recipient, TOKEN_ID, 1, obj.data);
+
+        return IERC721Receiver.onERC721Received.selector;
+    }
+
+    /**
+     * Deposit and wrap ERC-721 tokens.
+     * @param _tokenIds The ERC-721 token ids to deposit.
+     * @param _recipient The recipient of the wrapped tokens.
+     * @param _data Data to pass to ERC-1155 receiver.
+     * @notice Users must first approve this contract address on the ERC-721 contract.
+     * @notice This function can wrap multiple ERC-721 tokens at once.
+     */
+    function deposit(uint256[] calldata _tokenIds, address _recipient, bytes calldata _data) external {
+        if (_recipient == address(0)) {
+            revert InvalidDepositRequest();
+        }
+
+        uint256 length = _tokenIds.length;
+        for (uint256 i; i < length;) {
+            // Intentionally unsafe transfer
+            token.transferFrom(msg.sender, address(this), _tokenIds[i]);
+            unchecked {
+                // Can never overflow
+                i++;
+            }
+        }
+        emit TokensDeposited(_tokenIds);
+        _mint(_recipient, TOKEN_ID, length, _data);
+    }
+
+    //
+    // Withdraw
+    //
+
+    /**
+     * Accepts wrapped ERC-1155 tokens to unwrap.
+     * @param _amount The amount of tokens being transferred.
+     * @param _data Additional data with no specified format.
+     * @return `bytes4(keccak256("onERC1155Received(address,address,uint256,uint256,bytes)`
+     */
+    function onERC1155Received(address, address, uint256, uint256 _amount, bytes calldata _data)
+        external
+        returns (bytes4)
+    {
+        if (msg.sender != address(this)) {
+            revert InvalidERC1155Received();
+        }
+        _withdraw(_amount, _data);
+
+        return IERC1155TokenReceiver.onERC1155Received.selector;
+    }
+
+    /**
+     * Accepts wrapped ERC-1155 tokens to unwrap.
+     * @param _ids The IDs of the tokens being transferred.
+     * @param _amounts The amounts of tokens being transferred.
+     * @param _data Additional data formatted as WithdrawRequestObj.
+     * @return `bytes4(keccak256("onERC1155BatchReceived(address,address,uint256[],uint256[],bytes)"))`
+     */
+    function onERC1155BatchReceived(
+        address,
+        address,
+        uint256[] calldata _ids,
+        uint256[] calldata _amounts,
+        bytes calldata _data
+    ) public returns (bytes4) {
+        if (msg.sender != address(this)) {
+            revert InvalidERC1155Received();
+        }
+
+        if (_ids.length != 1) {
+            revert InvalidERC1155Received();
```

Comment:
It could be called with multiple token ids if the amounts are 0. 
```
safeBatchTransferFrom(_from, _to, [1, 2, 3], [0, 0, 0])
```

We could override the transfer functions and check for id length there instead. That would be more gas efficient too

Reply:
^ In progress. Blocked by https://github.com/0xsequence/erc-1155/pull/101